File: atcoder/abc/030/d.rb
class Node
    attr_accessor :here,:to
    def initialize
        @here = -1
        @to = -1
    end
end

n,a = gets.chomp.split.map(&:to_i)

k = gets.chomp.to_i

dict= Array.new(n+1)

for i in 1..n
    dict[i] = Node.new
end

v=[]

aa = gets.chomp.split.map(&:to_i)

for i in 1..n 
    dict[i].to = aa[i-1]
end

b=0

while true
    b = dict[a].to
    break if dict[b].here != -1
    v << b
    dict[b].here = v.size-1
    a=b
end

k -= dict[b].here
mod = v.size-dict[b].here

puts v[k%mod+dict[b].here-1]

File: atcoder/abc/030/d.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

struct Node{
    int here;
    int to;
};

int main(){
    ll N,a,k;
    vector<int> v;
    cin >> N >> a >> k;
    vector<Node> dict(N+1);

    for(int i=1;i<=N;i++){
        cin >> dict[i].to;
        dict[i].here=-1;
    }

    int b;
    while(true){
        b = dict[a].to;
        if(dict[b].here != -1){
            break;
        }else{
            v.push_back(b);
            dict[b].here=v.size()-1;
            a=b;
        }
    }

    k -= dict[b].here;
    ll mod = v.size()-dict[b].here;

    cout << v[k%mod+dict[b].here-1] << endl;
}
